File: Java-CodeServer/src/main/resources/js/prompto/runtime/Interpreter.js
var SyntaxError = require("../error/SyntaxError").SyntaxError;
var DictType = require("../type/DictType").DictType;
var TextType = require("../type/TextType").TextType;
var ArgumentAssignment = require("../grammar/ArgumentAssignment").ArgumentAssignment;
var ArgumentAssignmentList = require("../grammar/ArgumentAssignmentList").ArgumentAssignmentList;
var UnresolvedArgument = require("../grammar/UnresolvedArgument").UnresolvedArgument;
var DictLiteral = require("../literal/DictLiteral").DictLiteral;
var MethodCall = require("../statement/MethodCall").MethodCall;
var MethodSelector = require("../expression/MethodSelector").MethodSelector;

var argsType = new DictType(TextType.instance);

function parseCmdLineArgs(cmdLineArgs) {
	try {
		var args = CmdLineParser.parse(cmdLineArgs);
		var valueArgs = {};
		var keys = Object.keys(args);
		for(var i=0;i<keys.length;i++) {
			var key = new Text(keys[i]);
			var value = new Text(args[keys[i]]);
			valueArgs[key] = value;
		}
		var dict = new Dictionary(valueArgs);
		return new ExpressionValue(argsType, dict);
	} catch(e) {
		// TODO
		return new DictLiteral();
	}
}

function buildAssignments(method, cmdLineArgs) {
	var assignments = new ArgumentAssignmentList();
	if(method.args.length==1) {
		var name = method.args[0].name;
		var value = parseCmdLineArgs(cmdLineArgs);
		assignments.add(new ArgumentAssignment(new UnresolvedArgument(name), value));
	}
	return assignments;
}


function locateMethod(context, methodName, cmdLineArgs) {
	var map = context.getRegisteredDeclaration(methodName);
	if(map==null) {
		throw new SyntaxError("Could not find a \"" + methodName + "\" method.");
	}
	return locateMethodInMap(map, cmdLineArgs);
}

function locateMethodInMap(map, cmdLineArgs) {
	if(cmdLineArgs==null || cmdLineArgs.length==0) {
		return locateMethodWithArgs(map);
	} else {
		return locateMethodWithArgs(map, new DictType(TextType.instance));
	}
}

function locateMethodWithArgs(map) {
	// try exact match first
	var keys = Object.keys(map.methods);
	for(var i=0;i<keys.length;i++) {
		var method = map.methods[keys[i]];
		if(identicalArguments(method.args, arguments)) {
			return method;
		}
	}
	// match Text{} argument, will pass null
	if(arguments.length==1) {
		for (var i = 0; i < keys.length; i++) {
			var method = map.methods[keys[i]];
			if (isSingleTextDictArgument(method.args)) {
				return method;
			}
		}
	}
	// match no argument, will ignore options
	for (var i = 0; i < keys.length; i++) {
		var method = map.methods[keys[i]];
		if(method.args.length==0) {
			return method;
		}
	}
	throw new SyntaxError("Could not find a compatible \"" + map.name + "\" method.");
}

function isSingleTextDictArgument(args) {
	if(args.length!=1) {
		return false;
	}
	var arg = args[0];
	var typ = arg.type ||null
	if(typ==null) {
		return false;
	}
	return typ.equals(argsType);
}

function identicalArguments(args, argTypes) {
	if(args.length!=argTypes.length-1) {
		return false;
	}
	for(var i=0;i<args.length;i++) {
		var typ = args[i].typ || null;
		if(typ==null) {
			return false;
		}
		if(!typ.equals(argTypes[i+1])) {
			return false;
		}
	}
	return true;
}

function Interpreter() {
}

Interpreter.interpret = function(context, methodName, cmdLineArgs) {
	try {
		var method = locateMethod(context, methodName, cmdLineArgs);
		var assignments = buildAssignments(method, cmdLineArgs);
		var call = new MethodCall(new MethodSelector(null, methodName),assignments);
		call.interpret(context);
	} finally {
		context.terminated();
	}
};


Interpreter.interpretTests = function(context) {
    for(var name in context.tests) {
        var test = context.tests[name];
        var local = context.newLocalContext();
        test.interpret(local);
    }
};

exports.Interpreter = Interpreter;

File: Java-Core/src/main/java/prompto/runtime/Interpreter.java
package prompto.runtime;

import java.util.Collection;
import java.util.Map;
import java.util.Map.Entry;

import prompto.argument.IArgument;
import prompto.argument.ITypedArgument;
import prompto.argument.UnresolvedArgument;
import prompto.declaration.IMethodDeclaration;
import prompto.declaration.TestMethodDeclaration;
import prompto.error.PromptoError;
import prompto.error.SyntaxError;
import prompto.expression.IExpression;
import prompto.expression.MethodSelector;
import prompto.grammar.ArgumentAssignment;
import prompto.grammar.ArgumentAssignmentList;
import prompto.grammar.ArgumentList;
import prompto.grammar.Identifier;
import prompto.intrinsic.PromptoDict;
import prompto.literal.DictLiteral;
import prompto.runtime.Context.MethodDeclarationMap;
import prompto.statement.MethodCall;
import prompto.type.DictType;
import prompto.type.IType;
import prompto.type.TextType;
import prompto.utils.CmdLineParser;
import prompto.value.Dictionary;
import prompto.value.ExpressionValue;
import prompto.value.IValue;
import prompto.value.Text;

public class Interpreter {
	
	private static IType argsType = new DictType(TextType.instance());
			
	private Interpreter() {
	}
	
	public static void interpretTests(Context context) throws PromptoError {
		Collection<TestMethodDeclaration> tests = context.getTests();
		for(TestMethodDeclaration test : tests) {
			Context local = context.newLocalContext();
			test.interpret(local);
		}
		
	}

	public static void interpretTest(Context context, Identifier testName, boolean lookInStore) throws PromptoError {
		TestMethodDeclaration test = context.getTest(testName, lookInStore);
		Context local = context.newLocalContext();
		test.interpret(local);
	}
	
	public static void interpretMainNoArgs(Context context) throws PromptoError {
		interpretMethod(context, new Identifier("main"), "");
	}
	
	public static void interpretMethod(Context context, Identifier methodName, String cmdLineArgs) throws PromptoError {
		try {
			IMethodDeclaration method = locateMethod(context, methodName, cmdLineArgs);
			ArgumentAssignmentList assignments = buildAssignments(method, cmdLineArgs);
			MethodCall call = new MethodCall(new MethodSelector(methodName),assignments);
			call.interpret(context);	
		} finally {
			context.terminated();
		}
	}
	
	public static void interpretScript(Context context, String cmdLineArgs) throws PromptoError {
	}

	private static ArgumentAssignmentList buildAssignments(IMethodDeclaration method, String cmdLineArgs) {
		ArgumentAssignmentList assignments = new ArgumentAssignmentList();
		if(method.getArguments().size()==1) {
			Identifier name = method.getArguments().getFirst().getId();
			IExpression value = parseCmdLineArgs(cmdLineArgs);
			assignments.add(new ArgumentAssignment(new UnresolvedArgument(name), value)); 
		}
		return assignments;
	}

	private static IExpression parseCmdLineArgs(String cmdLineArgs) {
		try {
			Map<String,String> args = CmdLineParser.parse(cmdLineArgs);
			PromptoDict<Text, IValue> valueArgs = new PromptoDict<Text, IValue>(true);
			for(Entry<String,String> entry : args.entrySet())
				valueArgs.put(new Text(entry.getKey()), new Text(entry.getValue()));
			valueArgs.setMutable(false);
			Dictionary dict = new Dictionary(TextType.instance(), valueArgs);
			return new ExpressionValue(argsType, dict);
		} catch(Exception e) {
			// TODO
			return new DictLiteral(false);
		}
	}

	 private static IMethodDeclaration locateMethod(Context context, Identifier methodName, String cmdLineArgs) {
		MethodDeclarationMap map = context.getRegisteredDeclaration(MethodDeclarationMap.class, methodName);
		if(map==null)
			throw new SyntaxError("Could not find a \"" + methodName + "\" method.");
		return locateMethod(map, cmdLineArgs);
	}
			
	private static IMethodDeclaration locateMethod(MethodDeclarationMap map, String cmdLineArgs) {
		if(cmdLineArgs==null)
			return locateMethod(map);
		else
			return locateMethod(map, new DictType(TextType.instance()));
	}

	private static IMethodDeclaration locateMethod(MethodDeclarationMap map, IType ... argTypes) {
		// try exact match first
		for(IMethodDeclaration method : map.values()) {
			if(identicalArguments(method.getArguments(), argTypes))
				return method;
		}
		// match Text{} argument, will pass null 
		if(argTypes.length==0) for(IMethodDeclaration method : map.values()) {
			if(isSingleTextDictArgument(method.getArguments()))
				return method;
		}
		// match no argument, will ignore options
		for(IMethodDeclaration method : map.values()) {
			if(method.getArguments().size()==0)
				return method;
		}
		throw new SyntaxError("Could not find a compatible \"" + map.getId() + "\" method.");
	}

	private static boolean isSingleTextDictArgument(ArgumentList arguments) {
		if(arguments.size()!=1)
			return false;
		IArgument arg = arguments.getFirst();
		if(!(arg instanceof ITypedArgument))
			return false;
		return ((ITypedArgument)arg).getType().equals(argsType);
	}

	private static boolean identicalArguments(ArgumentList arguments, IType[] argTypes) {
		if(arguments.size()!=argTypes.length)
			return false;
		int idx = 0;
		for(IArgument argument : arguments) {
			if(!(argument instanceof ITypedArgument))
				return false;
			IType argType = argTypes[idx++];
			if(!argType.equals(((ITypedArgument)argument).getType()))
				return false;
		}
		return true;
	}

	
	
	
}
